File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/memory/search/AgentScopeFilter.kt
/**
 * AgentScopeFilter — tag-based KB isolation per agent role.
 * Each agent role has a configurable tag set. Search results are
 * filtered to only include entries matching the agent's tags.
 * Untagged entries remain visible to all agents (backward compatible).
 */
package com.codeintel.memory.search

import com.codeintel.memory.models.KnowledgeSearchResult
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import java.sql.Connection

/** Scope configuration for an agent role. */
data class AgentScope(val role: String, val tags: List<String>)

class AgentScopeFilter(private val conn: Connection) {

    private val cache = mutableMapOf<String, List<String>>()
    private val json = Json { ignoreUnknownKeys = true }

    init { loadCache() }

    /** Get scope configuration for an agent role. */
    fun getScope(agentRole: String): AgentScope? {
        val tags = cache[agentRole.uppercase()] ?: return null
        return AgentScope(agentRole.uppercase(), tags)
    }

    /** Filter search results by agent's tag set. */
    fun filter(results: List<KnowledgeSearchResult>, agentRole: String): List<KnowledgeSearchResult> {
        val scope = getScope(agentRole) ?: return results
        return results.filter { r ->
            val entryTags = parseTags(r.entry.tags)
            entryTags.isEmpty() || entryTags.any { it in scope.tags }
        }
    }

    /** Update scope configuration for a role. */
    fun updateScope(agentRole: String, tags: List<String>) {
        val role = agentRole.uppercase()
        val tagJson = "[" + tags.joinToString(",") { "\"$it\"" } + "]"
        conn.prepareStatement("""
            INSERT INTO agent_scope_config (agent_role, tag_set, updated_at)
            VALUES (?, ?, datetime('now'))
            ON CONFLICT(agent_role) DO UPDATE SET
              tag_set = excluded.tag_set, updated_at = datetime('now')
        """.trimIndent()).use { stmt ->
            stmt.setString(1, role)
            stmt.setString(2, tagJson)
            stmt.executeUpdate()
        }
        cache[role] = tags
    }

    private fun loadCache() {
        cache.clear()
        try {
            conn.createStatement().use { stmt ->
                stmt.executeQuery("SELECT agent_role, tag_set FROM agent_scope_config").use { rs ->
                    while (rs.next()) {
                        val role = rs.getString("agent_role")
                        val tagSet = rs.getString("tag_set")
                        val parsed = json.parseToJsonElement(tagSet).jsonArray
                        val tags = parsed.map { it.jsonPrimitive.content }
                        cache[role] = tags
                    }
                }
            }
        } catch (_: Exception) {
            // Table may not exist yet (pre-migration)
        }
    }
}

/** Parse comma-separated tags string into list. */
private fun parseTags(tags: String?): List<String> {
    if (tags.isNullOrBlank()) return emptyList()
    return tags.split(",").map { it.trim().lowercase() }.filter { it.isNotEmpty() }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/agent_scope_filter.py
"""AgentScopeFilter — tag-based KB isolation per agent role.

Each agent role has a configurable tag set. Search results are
filtered to only include entries matching the agent's tags.
Untagged entries remain visible to all agents (backward compatible).
"""

import json
import logging
import sqlite3
from dataclasses import dataclass
from typing import Any

logger = logging.getLogger(__name__)


@dataclass
class AgentScope:
    """Scope configuration for an agent role."""

    role: str
    tags: list[str]


class AgentScopeFilter:
    """Tag-based KB isolation per agent role."""

    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn
        self._cache: dict[str, list[str]] = {}
        self._load_cache()

    def get_scope(self, agent_role: str) -> AgentScope | None:
        """Get scope configuration for an agent role."""
        tags = self._cache.get(agent_role.upper())
        if tags is None:
            return None
        return AgentScope(role=agent_role.upper(), tags=tags)

    def filter(self, results: list[dict[str, Any]], agent_role: str) -> list[dict[str, Any]]:
        """Filter search results by agent's tag set."""
        scope = self.get_scope(agent_role)
        if scope is None:
            return results

        filtered = []
        for r in results:
            entry_tags = _parse_tags(r.get("entry", {}).get("tags"))
            if not entry_tags or any(t in scope.tags for t in entry_tags):
                filtered.append(r)
        return filtered

    def update_scope(self, agent_role: str, tags: list[str]) -> None:
        """Update scope configuration for a role."""
        role = agent_role.upper()
        tag_json = json.dumps(tags)
        cursor = self._conn.cursor()
        cursor.execute(
            """INSERT INTO agent_scope_config (agent_role, tag_set, updated_at)
               VALUES (?, ?, datetime('now'))
               ON CONFLICT(agent_role) DO UPDATE SET
                 tag_set = excluded.tag_set, updated_at = datetime('now')""",
            (role, tag_json),
        )
        self._conn.commit()
        self._cache[role] = tags

    def _load_cache(self) -> None:
        self._cache.clear()
        try:
            cursor = self._conn.cursor()
            cursor.execute("SELECT agent_role, tag_set FROM agent_scope_config")
            for row in cursor.fetchall():
                self._cache[row[0]] = json.loads(row[1])
        except sqlite3.OperationalError:
            # Table may not exist yet (pre-migration)
            logger.debug("agent_scope_config table not found, skipping cache load")


def _parse_tags(tags: str | None) -> list[str]:
    """Parse comma-separated tags string into list."""
    if not tags:
        return []
    return [t.strip().lower() for t in tags.split(",") if t.strip()]
